File: data/two-pointers/valid-palindrome.cpp
#include <iostream>
#include <cctype>  // For isalnum() and tolower()
using namespace std;

bool isPalindrome(string s) {
    int left = 0, right = s.length() - 1;  // Initialize two pointers

    while (left < right) {
        while (left < right && !isalnum(s[left])) left++;  // Ignore non-alphanumeric characters
        while (left < right && !isalnum(s[right])) right--;  // Ignore non-alphanumeric characters

        if (tolower(s[left]) != tolower(s[right])) return false;  // Case-insensitive comparison

        left++;
        right--;
    }

    return true;  // If no mismatches, it's a palindrome
}

// Example Usage
int main() {
    cout << boolalpha;  // Print boolean values as true/false
    cout << isPalindrome("A man, a plan, a canal: Panama") << endl;  // Output: true
    cout << isPalindrome("race a car") << endl;  // Output: false
    cout << isPalindrome(" ") << endl;  // Output: true
    return 0;
}

File: data/two-pointers/valid-palindrome.py
def isPalindrome(s):
    """
    Check if a given string is a palindrome, considering only alphanumeric characters and ignoring cases.

    :param s: str - Input string
    :return: bool - True if palindrome, False otherwise
    """
    left, right = 0, len(s) - 1  # Initialize two pointers

    while left < right:
        while left < right and not s[left].isalnum():  # Ignore non-alphanumeric characters
            left += 1
        while left < right and not s[right].isalnum():  # Ignore non-alphanumeric characters
            right -= 1
        
        if s[left].lower() != s[right].lower():  # Case-insensitive comparison
            return False
        
        left += 1  # Move left pointer forward
        right -= 1  # Move right pointer backward

    return True  # If no mismatches, it's a palindrome

# Example Usage
print(isPalindrome("A man, a plan, a canal: Panama"))  # Output: True
print(isPalindrome("race a car"))  # Output: False
print(isPalindrome(" "))  # Output: True (An empty string is a palindrome)
